bpf/lib/lb.h

```diff
@@ -51,16 +56,75 @@ __BPF_MAP(cilium_lb4_services, BPF_MAP_TYPE_HASH, 0,
 	  sizeof(struct lb4_key), sizeof(struct lb4_service),
 	  PIN_GLOBAL_NS, CILIUM_LB_MAP_MAX_ENTRIES);
 
+__BPF_MAP(cilium_lb4_rr_seq, BPF_MAP_TYPE_HASH, 0,
+	  sizeof(struct lb4_key), sizeof(struct lb_sequence),
+	  PIN_GLOBAL_NS, CILIUM_LB_MAP_MAX_FE);
+
 #define REV_NAT_F_TUPLE_SADDR 1
+#ifdef LB_DEBUG
+#define cilium_trace_lb cilium_trace
+#else
+#define cilium_trace_lb(a, b, c, d)
+#endif
+
+static inline int lb_next_rr(struct __sk_buff *skb,
+			     struct lb_sequence *seq,
+			     __be16 hash)
+{
+	int slave = 0;
+	__u8 offset = hash % seq->count;
+
+	if (offset < LB_RR_MAX_SEQ) {
+		/* Slave 0 is reserved for the master slot */
+		slave = seq->idx[offset] + 1;
+		cilium_trace(skb, DBG_RR_SLAVE_SEL, hash, slave);
+	}
+
+	return slave;
+}
+
+static inline int lb6_select_slave(struct __sk_buff *skb,
+				   struct lb6_key *key,
+				   __u16 count, __u16 weight)
+{
+	__be16 hash = get_hash_recalc(skb);
+	struct lb_sequence *seq;
+	int slave = 0;
+
+	if (weight) {
+		seq = map_lookup_elem(&cilium_lb6_rr_seq, key);
+		if (seq && seq->count != 0)
+			slave = lb_next_rr(skb, seq, hash);
+	}
+
+	if (slave == 0) {
+		/* Slave 0 is reserved for the master slot */
```

Comment:
master just means the front-end. So we do two lookups in same table. One to get the front-end from the packet when slave index will be 0, and next to get the backend with the appropriate slave index > 0. So if the RR is not configured we fall back to hash based.